pageserver/src/tenant/disk_btree.rs

```diff
@@ -525,19 +525,128 @@ where
     }
 }
 
-pub struct DiskBtreeIterator<'a> {
+pub trait DiskBtreeIterator {
+    type Item;
+    async fn next(&mut self) -> Option<std::result::Result<Self::Item, DiskBtreeError>>;
+    fn map<B, F>(self, f: F) -> MapDiskBtreeIter<Self, F>
+    where
+        Self: Sized,
+        F: FnMut(Self::Item) -> B,
+    {
+        MapDiskBtreeIter::new(self, f)
+    }
+
+    fn filter<P>(self, predicate: P) -> FilterDiskBtreeIter<Self, P>
+    where
+        Self: Sized,
+        P: FnMut(&Self::Item) -> bool,
+    {
+        FilterDiskBtreeIter::new(self, predicate)
+    }
+}
```

Comment:
This seems to me like this could be a `futures::future::Stream`, it would give you the (async) map and sync filter(?) right away without writing these up.